_posts/2022-11-11-functional-crystal.md

```diff
@@ -0,0 +1,431 @@
+
+<!-- If we want to be able to alter the value of `y`, we need to pass it by reference. In Crystal this is doable, but to do it right you need to _box_ it. We won't dicuss this -->
+
+Consider now the following code:
+
+```cr
+def g(arr : Array(Int32))
+  arr += [1]
+  arr = [] of Int32
+end
+
+z = [0]
+g z
+p z
+```
+
+What does it print?
+
+<br/>
+<br/>
+<center>
+Space purposefully left blank to let you think.
+</center>
+<br/>
+<br/>
+
+The answer is in the `Array`'s type: it's a `class`, so it's passed _by reference_. That is, changes performed in the array within the function affects `z`. Now, _the reference_ of the array itself (where `z` is located in the memory) is passed by value. Essentially, this means that the array itself is **not** replaced when assigning the empty array to `arr`. Therefore, the answer is `[0, 1]`.
+
+> An instance of a class is passed _by reference_, but the reference itself is passed _by value_.
+
+Crystal distinguishes which objects to pass by reference and which by value with two classes: [Reference](https://crystal-lang.org/api/1.6.2/Reference.html) and [Value](https://crystal-lang.org/api/1.6.2/Value.html). Those classes that inherits from `Reference` are passed by reference, and those that inherit from `Value`, by value.
+
+### Structs, immutable objects
+
+As mentioned previously, instances of `struct`s are passed by value. The following example about 2D points in space shows how structs work:
+
+```cr
+struct Point
+  getter x : Int32, y : Int32
+
+  def initialize(@x, @y); end
+end
+
+def translate_x(point : Point, offset : Int32)
+  point.x += offset
+  point
+end
+
+zero = Point.new 0, 0
+ten = translate_x zero, 10
+p zero, ten # => Point(@x=0, @y=0)  Point(@x=10, @y=0)
+```
+
+Note what's happening here: the function `translate_x` alters the value of the `x` component, and this change works locally: the _new copy_ of the point that is returned has its value altered. See how important immutability is: we _know_ that if we do not change `zero` locally, it will always have the value `(0, 0)`.
+
+Exercise 1: replace `struct` with `class` and see what happens.
+
+Exercise 2: make `translate_x` be an instance method of `struct`, that operates on `@x` of `self` instead of receiving the point. Can you explain what happens?
+
+### Little diversion: the `record` macro
+
+An alternative implementation of the `Point` struct is using the macro [`record`](https://crystal-lang.org/api/1.6.2/toplevel.html#record%28name%2C%2Aproperties%29-macro). This macro further enhances the experience of working with immutable structs.
+
+```cr
+record Point, x : Int32 = 0, y : Int32 = 0
+```
+
+`record` generates an identical struct as above, together with a handy method `#copy_with` that allows to return a copy of the struct with some given instance variables modified. For instance, we can solve Exercise 2 above defining the `translate_x` method directly when defining `Point`:
```

Comment:
If the above example uses `property` instead of `getter` then theseare not identical structs, as only the one above is mutable